File: herald-admin-go/transport.go
package herald

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type httpTransport struct {
	baseURL string
	token   string
	client  *http.Client
}

func newTransport(baseURL, token string) *httpTransport {
	return &httpTransport{
		baseURL: strings.TrimRight(baseURL, "/"),
		token:   token,
		client:  &http.Client{},
	}
}

func (t *httpTransport) request(ctx context.Context, method, path string, body any) (json.RawMessage, error) {
	var bodyReader io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("marshal body: %w", err)
		}
		bodyReader = bytes.NewReader(b)
	}

	req, err := http.NewRequestWithContext(ctx, method, t.baseURL+path, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+t.token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("http request: %w", err)
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode >= 400 {
		httpCodes := map[int]string{
			400: "BAD_REQUEST", 401: "UNAUTHORIZED", 403: "FORBIDDEN",
			404: "NOT_FOUND", 409: "CONFLICT", 429: "RATE_LIMITED",
			500: "INTERNAL", 503: "UNAVAILABLE",
		}
		code := "INTERNAL"
		if c, ok := httpCodes[resp.StatusCode]; ok {
			code = c
		}
		msg := fmt.Sprintf("HTTP %d", resp.StatusCode)
		var errResp struct {
			Error string `json:"error"`
		}
		if json.Unmarshal(data, &errResp) == nil && errResp.Error != "" {
			msg = errResp.Error
		} else if len(data) > 0 {
			msg = string(data)
		}
		return nil, &HeraldError{Code: code, Message: msg, Status: resp.StatusCode}
	}

	if resp.StatusCode == 204 {
		return nil, nil
	}

	return json.RawMessage(data), nil
}

File: herald-admin-python/herald_admin/transport.py
from __future__ import annotations

import json
import urllib.request
import urllib.error
import urllib.parse
from typing import Any

from .errors import HeraldError


class HttpTransport:
    def __init__(self, base_url: str, token: str, timeout: int = 30) -> None:
        self._base_url = base_url.rstrip("/")
        self._token = token
        self._timeout = timeout

    def request(self, method: str, path: str, body: Any = None) -> Any:
        url = f"{self._base_url}{path}"
        data = json.dumps(body).encode() if body is not None else None

        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Authorization", f"Bearer {self._token}")
        if data is not None:
            req.add_header("Content-Type", "application/json")

        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                if resp.status == 204:
                    return None
                raw = resp.read()
                if not raw:
                    return None
                return json.loads(raw)
        except urllib.error.HTTPError as e:
            code = "INTERNAL"
            message = f"HTTP {e.code}"
            try:
                body_data = json.loads(e.read())
                if "error" in body_data:
                    message = body_data["error"]
            except Exception:
                pass
            if e.code == 401:
                code = "UNAUTHORIZED"
            elif e.code == 404:
                code = "NOT_FOUND"
            raise HeraldError(code, message, e.code) from None
